### PR TITLE
Clean self-signed ca

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/bootstrap/istio_ca.go
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/istio_ca.go:444]: meaningless

File: security/pkg/pki/ca/ca.go
Comment (hzxuzhonghu) [inline security/pkg/pki/ca/ca.go:193]: improve resilience

File: security/pkg/pki/ca/ca.go
Comment (howardjohn) [inline security/pkg/pki/ca/ca.go:129]: This is a big change? As far as I can tell we went from not retrying to retrying. Can you expand on why we need this? What use cases is it fixing?

**Review by howardjohn — COMMENTED** [review]
This handling is pretty sensitive and complex. I think there were many PRs trying to do this before that all had bugs and weren't merged as a result.

Can we add tests of the error cases when we have multiple Istiods trying to read/write the secret concurrently? Specifically we want to ensure they all end up with a consistent state in the end.

File: security/pkg/pki/ca/ca.go
Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca.go:125]: CUrrent no retry at all because -1 is passed in

File: security/pkg/pki/ca/ca.go
Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca.go:121]: maybe i can try

File: security/pkg/pki/ca/ca.go
Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca.go:129]: It is to keep the original intention, no strong preference. If we all agree, I can make it fail fast.

File: security/pkg/k8s/secret/secret_util.go
Comment (hzxuzhonghu) [reply security/pkg/k8s/secret/secret_util.go:34]: It is never announced publicly. And from istio, i can not find any use of it.

Comment (lei-tang) [reply security/pkg/k8s/secret/secret_util.go:34]: To reduce the risk of breaking existing users, consider split the removal of "istio.io/service-account.name" to a separate PR.

**Comment by lei-tang** [comment]
I agree with John's concerns. https://github.com/istio/istio/pull/40657 is an example that tried to update this code and its review comments explain why concurrent write and read CA secret and bundles can cause inconsistency problems. The existing code is is simple, easy to reason its behavior under concurrency, and battle tested.

File: security/pkg/pki/ca/ca_test.go
Comment (lei-tang) [inline security/pkg/pki/ca/ca_test.go:313]: Can you also check the consistency of CA secret and CA bundle?

Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca_test.go:313]: Sure, will update

Comment (lei-tang) [reply security/pkg/pki/ca/ca_test.go:313]: For example, checking that the CA secret is consistent with the CA bundle.

Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca_test.go]: done

**Comment by hzxuzhonghu** [comment]
@howardjohn Added a test, and have figured out the original issue with @lei-tang offline. It was not returning create secret error.

FYI https://github.com/istio/istio/issues/14512

File: security/pkg/k8s/secret/secret_util.go
Comment (howardjohn) [inline security/pkg/k8s/secret/secret_util.go:43]: I think someone had concerns with this. IMO its fine, this is dead code since saName is always `""`

File: security/pkg/pki/ca/ca.go
Comment (howardjohn) [inline security/pkg/pki/ca/ca.go:186]: `Abort` is not accurate any more, right? Now we retry?

Comment (howardjohn) [reply security/pkg/pki/ca/ca.go:186]: very nit: may be nice to specifically handle the IsAlreadyExist error with a non `Errorf` level log? Since it is expected case

File: security/pkg/pki/ca/ca.go
Comment (howardjohn) [inline security/pkg/pki/ca/ca.go:148]:
```suggestion
			pkiCaLog.Infof("Load signing key and cert from existing secret %s: %s", caSecret.Namespace, caSecret.Name)
```

small nit

**Review by howardjohn — COMMENTED** [review]
LGTM code is much simpler and test is great. will leave in case Lei has other comments

File: security/pkg/pki/ca/ca_test.go
Comment (shankgan) [inline security/pkg/pki/ca/ca_test.go:231]: are we removing any expected error from this test case?

File: security/pkg/k8s/secret/secret_util.go
Comment (costinm) [reply security/pkg/k8s/secret/secret_util.go:34]: Is this the annotation on Service ?

File: pilot/pkg/bootstrap/istio_ca.go
Comment (costinm) [reply pilot/pkg/bootstrap/istio_ca.go:444]: If you really want to clean - replacing the 20 or so params with a struct would make things far more readable.

**Comment by costinm** [comment]
I think it looks good, but still feel a bit worried...

**Comment by lei-tang** [comment]
> I think it looks good, but still feel a bit worried...

Using a flag to enable the new behavior may reduce the risks.

**Comment by costinm** [comment]
But it defeats the purpose of the cleanup.

I think it's reasonable to have a flag to indicate if Citadel is required -
triggering fail fast. But I think we have such a flag already.

File: security/pkg/pki/ca/ca.go
Comment (hzxuzhonghu) [reply security/pkg/pki/ca/ca.go:186]: good catch

**Comment by hzxuzhonghu** [comment]
ping @istio/wg-security-maintainers

**Review by lei-tang — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
@howardjohn can you give a final look

File: security/pkg/k8s/secret/secret_util.go
Comment (hzxuzhonghu) [reply security/pkg/k8s/secret/secret_util.go:34]: Actually dead code now

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
